java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -158,198 +106,162 @@ class EntityHandleManager(private val handleManager: HandleManager) {
-    val storageHandle: SingletonHandle<RawEntity>,
-    sender: Sender
-) : HandleEventBase<T?, ReadSingletonHandle<T>>(), ReadSingletonHandle<T> {
+/** Wraps [SingletonHandle] and makes it suitable for use in the SDK with [Entity]s. */
+class SingletonHandleAdapter<T : Entity>(
+    mode: HandleMode,
+    name: String,
+    private val entitySpec: EntitySpec<T>,
+    private val storageHandle: SingletonHandle<RawEntity>,
+    private val idGenerator: Id.Generator,
+    private val sender: Sender
+) : HandleAdapter(mode, name), ReadWriteSingletonHandle<T> {
     init {
-        storageHandle.callback = SingletonSenderCallbackAdapter(
-            this::fetch,
-            this::fireUpdate,
-            this::fireSync,
-            this::fireDesync,
-            sender
+        storageHandle.callback = LambdaCallbacks(
+            onSync = this::fireSync,
+            onDesync = this::fireDesync,
+            onUpdate = this::fireUpdate
         )
     }
 
-    override val name: String
-        get() = handleName
-
-    override suspend fun fetch(): T? = storageHandle.fetch()?.let {
-        rawEntity -> entitySpec.deserialize(rawEntity)
-    }
-}
-
-internal class WriteSingletonHandleImpl<T : Entity>(
-    val handleName: String,
-    private val storageHandle: SingletonHandle<RawEntity>,
-    private val idGenerator: Id.Generator
-) : WriteSingletonHandle<T> {
-    override val name: String
-        get() = handleName
+    private val onUpdateActions: MutableList<(T?) -> Unit> = mutableListOf()
 
     override suspend fun store(entity: T) {
+        checkCanWrite()
         storageHandle.store(
-            entity.ensureIdentified(idGenerator, handleName).serialize()
+            entity.ensureIdentified(idGenerator, name).serialize()
         )
     }
 
     override suspend fun clear() {
+        checkCanWrite()
         storageHandle.clear()
     }
-}
 
-internal class ReadWriteSingletonHandleImpl<T : Entity>(
-    entitySpec: EntitySpec<T>,
-    handleName: String,
-    storageHandle: SingletonHandle<RawEntity>,
-    idGenerator: Id.Generator,
-    sender: Sender,
-    private val writableSingleton: WriteSingletonHandleImpl<T> = WriteSingletonHandleImpl(
-        handleName,
-        storageHandle,
-        idGenerator
-    )
-) : ReadWriteSingletonHandle<T>,
-    ReadSingletonHandleImpl<T>(entitySpec, handleName, storageHandle, sender),
-    WriteSingletonHandle<T> by writableSingleton {
-    override val name: String
-        get() = writableSingleton.name
+    override suspend fun fetch(): T? {
+        checkCanRead()
+        return storageHandle.fetch()?.let { rawEntity ->
+            entitySpec.deserialize(rawEntity)
+        }
+    }
+
+    override fun onUpdate(action: (T?) -> Unit) {
+        checkCanRead()
+        onUpdateActions.add(action)
+    }
+
+    private fun fireUpdate() {
+        checkCanRead()
+        sender {
+            val value = fetch()
+            onUpdateActions.forEach { it(value) }
```

Comment:
Now that I think about it, it might be a good idea to surround these with a try/catch + log, so that particles throwing exceptions don't propagate all the way up the storage stack. I don't know if these are protected higher up, but if the call stack is straight from onMessage(), an exception might interrupt something in StorageProxy that leaves it in an inconsistent state?

Might be worth investigating.

Reply:
Actually, the sender can just add exception catching if desired.